src/software/geom/algorithms/end_in_obstacle_sample_test.cpp

```diff
@@ -0,0 +1,196 @@
+{
+    Field field              = world.field();
+    Rectangle navigable_area = field.fieldLines();
+    std::vector<ObstaclePtr> obstacles;
+
+    std::vector<ObstaclePtr> field_boundary =
+        obstacle_factory.createStaticObstaclesFromMotionConstraint(
+            TbotsProto::MotionConstraint::AVOID_FIELD_BOUNDARY_ZONE, field);
+    obstacles.insert(obstacles.end(), field_boundary.begin(), field_boundary.end());
+
+    Point destination(4.9, 2);
+    std::optional<Point> end_point =
+        endInObstacleSample(obstacles, destination, navigable_area);
+
+    if (end_point.has_value())
+    {
+        double min_distance = MAX_ALLOWABLE_SAMPLE_ERROR + 1;
+        for (auto const &obstacle : obstacles)
+        {
+            double point_distance_from_obstacle =
+                distance(end_point.value(), obstacle->closestPoint(end_point.value()));
+            min_distance = std::min(min_distance, point_distance_from_obstacle);
+            // check that the returned point is not inside an obstacle
+            ASSERT_FALSE(obstacle->contains(end_point.value()));
+        }
+        // check that returned point is not too far from the nearest obstacle
+        if (min_distance > MAX_ALLOWABLE_SAMPLE_ERROR)
+        {
+            std::cout << "end_point" << end_point.value() << '\n';
+            FAIL() << "Returned point is too far from destination";
+        }
+    }
+    else
+    {
+        FAIL();
+    }
+}
+
+TEST_F(TestEndInObstacleSampler, test_end_in_defense_area)
+{
+    Field field              = world.field();
+    Rectangle navigable_area = field.fieldBoundary();
+    std::vector<ObstaclePtr> obstacles;
+
+    std::vector<ObstaclePtr> friendly_defense_area =
+        obstacle_factory.createStaticObstaclesFromMotionConstraint(
+            TbotsProto::MotionConstraint::FRIENDLY_DEFENSE_AREA, field);
+    obstacles.insert(obstacles.end(), friendly_defense_area.begin(),
+                     friendly_defense_area.end());
+
+    Point destination(-4, 0.5);
+    std::optional<Point> end_point =
+        endInObstacleSample(obstacles, destination, navigable_area);
+
+    if (end_point.has_value())
+    {
+        double min_distance = MAX_ALLOWABLE_SAMPLE_ERROR + 1;
+        for (auto const &obstacle : obstacles)
+        {
+            double point_distance_from_obstacle =
+                distance(end_point.value(), obstacle->closestPoint(end_point.value()));
+            min_distance = std::min(min_distance, point_distance_from_obstacle);
+            // check that the returned point is not inside an obstacle
+            ASSERT_FALSE(obstacle->contains(end_point.value()));
+        }
+        // check that returned point is not too far from the nearest obstacle
+        if (min_distance > MAX_ALLOWABLE_SAMPLE_ERROR)
+        {
+            std::cout << "end_point" << end_point.value() << '\n';
+            FAIL() << "Returned point is too far from destination";
+        }
+    }
+    else
+    {
+        FAIL();
+    }
```

Comment:
nit: This validation seems to be repeated across the tests, could you refactor this to a method in the test class above that takes in the point and a set up obstacles?